explorer/interpreter/type_checker.cpp

```diff
@@ -1615,11 +1621,13 @@ class TypeChecker::ConstraintTypeBuilder {
       CARBON_ASSIGN_OR_RETURN(
           const auto* interface,
           type_checker.SubstituteCast<InterfaceType>(
-              local_bindings, &rewrite_constraint.constant->interface()));
+              local_bindings, &rewrite_constraint.constant->interface(),
+              rewrite_constraint.constant->constant().source_loc()));
```

Comment:
I don't think this is a useful location: this is where the associated constant was declared (eg, the `let ...` binding inside an interface), which might be far distant from the place that triggered a substitution to be performed. I think the location that you used above -- the location of the new constraint type's self binding -- is a good location for all the `Substitute` calls in this class (and for the `RebuildValue` calls -- see comment on that below).

It'd be nice to factor out the `constraint->self_binding()->source_loc()` expression into a member function (maybe `ConstraintTypeBuilder::source_loc()`) since it seems like it's going to come up a lot.